agent/service.go

```diff
@@ -15,165 +15,193 @@ package main
 
 import (
 	"context"
-	"io"
-	"io/ioutil"
 	"net"
-	"os"
 	"path/filepath"
-	"syscall"
+	"runtime/debug"
+	"time"
 
 	"github.com/containerd/containerd/cio"
+	"github.com/containerd/containerd/events/exchange"
 	"github.com/containerd/containerd/log"
 	"github.com/containerd/containerd/namespaces"
-	"github.com/containerd/containerd/runtime/v2/shim"
-	shimapi "github.com/containerd/containerd/runtime/v2/task"
-	"github.com/containerd/fifo"
+	"github.com/containerd/containerd/runtime/v2/runc"
+	taskAPI "github.com/containerd/containerd/runtime/v2/task"
 	"github.com/gogo/protobuf/types"
 	"github.com/mdlayher/vsock"
+	"github.com/pkg/errors"
 	"github.com/sirupsen/logrus"
 
-	"github.com/firecracker-microvm/firecracker-containerd/internal"
+	"github.com/firecracker-microvm/firecracker-containerd/internal/bundle"
+	"github.com/firecracker-microvm/firecracker-containerd/internal/vm"
 	"github.com/firecracker-microvm/firecracker-containerd/proto"
 )
 
 const (
-	bundleMountPath  = "/container"
-	defaultStdioPath = "/container/fifo"
+	containerRootDir = "/container"
 )
 
 // TaskService represents inner shim wrapper over runc in order to:
 // - Add default namespace to ctx as it's not passed by ttrpc over vsock
 // - Add debug logging to simplify debugging
 // - Make place for future extensions as needed
 type TaskService struct {
-	runc    shim.Shim
-	cancels []context.CancelFunc
-	io      *cio.FIFOSet
+	vm.TaskManager
+
+	eventExchange *exchange.Exchange
+	cancel        func()
 }
 
 // NewTaskService creates new runc shim wrapper
-func NewTaskService(runc shim.Shim, cancel context.CancelFunc) shimapi.TaskService {
+func NewTaskService(eventExchange *exchange.Exchange, cancel context.CancelFunc) taskAPI.TaskService {
 	return &TaskService{
-		runc:    runc,
-		cancels: []context.CancelFunc{cancel},
+		eventExchange: eventExchange,
+		cancel:        cancel,
 	}
 }
 
-// Create creates a new initial process and container using runc
-func (ts *TaskService) Create(ctx context.Context, req *shimapi.CreateTaskRequest) (*shimapi.CreateTaskResponse, error) {
-	log.G(ctx).WithFields(logrus.Fields{"id": req.ID, "bundle": req.Bundle}).Info("create")
+func logPanic(logger *logrus.Entry) {
```

Comment:
nit: looks like we recover here in the first place and logging is a consequence. I'd propose `recoverPanic(ctx context.Context)`

Reply:
Well we recover, write the log, but then Fatal (which should exit the process), so we don't keep on running.

I can change the name to `logPanicAndDie` or something like that

Reply:
I didn't notice `Fatal`. Do we really want to panic Agent? As I understand there no restart mechanism, so we'll lose all containers in a VM.

Reply:
Right now I think yeah it's best to prefer just crashing. Panics should really only be thrown if something so terrible has happened that the program is better off crashing (i.e. a nil pointer exception that, if ignored, could result in inconsistent or corrupted state).

In the future if we find that we are getting crashes in situations where we are better off still running then we can update this function in the future to not Fatal.